test/Hanoi.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE FlexibleInstances  #-}
+
+------------------------------------------------------------------------
+
+-- Solution to the famous Tower of Hanoi puzzle using tools for state
+-- machine property-based testing.
+--
+-- The puzzle is to move N discs of different sizes from one peg to
+-- another, with one auxiliary peg and a restriction that no disc may ever
+-- be placed on top of a smaller disc. Only one disc can be moved at a time.
+
+------------------------------------------------------------------------
+
+module Hanoi
+  ( Command(..)
+  , Model(..)
+  , initModel
+  , transitions
+  , prop_hanoi
+  ) where
+
+import Data.Kind
+         (Type)
+import Data.Array
+import Data.Maybe
+import Data.TreeDiff.Expr ()
+import GHC.Generics
+         (Generic, Generic1)
+import Prelude
+import Test.QuickCheck
+         (Gen, Property, (===), Arbitrary (arbitrary), choose, suchThat)
+import Test.QuickCheck.Monadic
+         (monadicIO)
+import Test.StateMachine
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+
+------------------------------------------------------------------------
+
+-- The model keeps track of which disc is on which peg
+
+newtype Model (r :: Type -> Type) = Model (Array Int [Int])
+  deriving (Show, Eq, Generic)
+
+-- There are 3 pegs, so the bounds are (0, 2)
+
+pegsBounds :: (Int,Int)
+pegsBounds = (0, 2)
+
+instance ToExpr (Model r) where
+  toExpr (Model a) = toExpr $ elems a
+
+initModel :: Int -> Model r
+initModel discs = Model $ listArray pegsBounds [[1..discs], [], []]
+
+-- Allowed action is to move one disc from the top of one peg to the top of another
+
+data Command (r :: Type -> Type) = Move (Int,Int)
+  deriving (Eq, Show, Generic1, Rank2.Functor, Rank2.Foldable, Rank2.Traversable, CommandNames)
+
+instance Arbitrary (Command r) where
+  arbitrary = do
+    x <- choose pegsBounds
+    y <- choose pegsBounds `suchThat` (/= x)
+    return $ Move (x,y)
+
+data Response (r :: Type -> Type) = Done
+  deriving (Show, Generic1, Rank2.Foldable)
+
+------------------------------------------------------------------------
+
+transitions :: Model r -> Command r -> Response r -> Model r
+transitions (Model pegs) (Move (from,to)) _ =
+  Model $ pegs // [(from, xs), (to, x:(pegs!to))]
+    where (x, xs) = (head pegs_from, tail pegs_from)
```

Comment:
This can be simplified as follows:
```haskell
where
  (x :: xs) = pegs ! from
```

Reply:
I am not happy with my version either, but `(x : xs) = ...` caused a warning about non-exhaustive pattern match - I think it was safe, because precondition guarantees `pegs ! from` won't be `[]`, but still ugly (and not sure if shrinking maintains preconditions anyway?).
I will go with version below if that's ok
```haskell
transitions m@(Model pegs) (Move (from,to)) _ = case pegs ! from of
  (x : xs) -> Model $ pegs // [(from, xs), (to, x:(pegs ! to))]
  _ -> m
```

Reply:
Shrinking preserves pre-conditions. Rather do `[] -> error "transition: impossible, due to preconditon" and just returning `m`.